src/collapse/common.rs

```diff
@@ -0,0 +1,794 @@
+    /// is present, this method **must** consume it (i.e. advance the provided
+    /// reader past it).
+    ///
+    /// This method also provides an opportunity to do processing of actual
+    /// stack data on the main thread before worker threads are spun up. For
+    /// an example of why this might be necessary, see `perf`, which can require
+    /// reading the first stack in order to know how to process the rest.
+    ///
+    /// If the format you are working with does not contain header information
+    /// or does not need any special, up-front processing, just have this method
+    /// return `Ok(())` immediately.
+    fn pre_process<R>(&mut self, reader: &mut R, occurrences: &mut Occurrences) -> io::Result<()>
+    where
+        R: io::BufRead;
+
+    /// The primary method.
+    ///
+    /// It receives a reader whose header has already been consumed (see above), as
+    /// well as a mutable reference to an `Occurences` instance (just a hashamp that
+    /// works across multiple threads). Implementators should parse the stack data
+    /// contained in the reader and write output to the provided `Occurrences` map.
+    ///
+    /// So that this method may be called multiple times, all internal
+    /// state contained in `self` (e.g. any stack buffers or the like) **must** be
+    /// reset before this method returns. Also, this method may **not** use threads.
+    fn collapse_single_threaded<R>(
+        &mut self,
+        reader: R,
+        occurrences: &mut Occurrences,
+    ) -> io::Result<()>
+    where
+        R: io::BufRead;
+
+    /// The following method is used to determine how to chunk up the input data in order to
+    /// send it off to worker threads, which **must** receive full stacks (as opposed to partial
+    /// stacks). Since what comprises a "stack" will vary from format to format, implementors for
+    /// a specific format need to implement this method, which encodes how we know when we've
+    /// reached the end of a stack.
+    ///
+    /// This method **must** return `true` if the provided line represents the end of a stack;
+    /// `false` otherwise.
+    ///
+    /// If your format requires more information than merely a line of the input data in order
+    /// to determine whether or not you are at the end of a stack, you can access information
+    /// stored on the `self` instance, which is also available to you in this method.
+    fn would_end_stack(&self, line: &[u8]) -> bool;
+
+    /// This method should return whether the implementation corresponds with
+    /// the given input string, i.e. if the input data matches the collapser.
+    ///
+    /// - `None` means "not sure -- need more input"
+    /// - `Some(true)` means "yes, this implementation should work with this string"
+    /// - `Some(false)` means "no, this implementation definitely won't work"
+    fn is_applicable_(&mut self, input: &str) -> Option<bool>;
```

Comment:
I still find it _so_ weird that this method is suffixed by a `_`. Why is that necessary?

Reply:
It's not necessary. I just pushed a commit where they are no longer there. I'm fine with either version. The reason why they were there in the first place is that these methods (`collapse` and `is_applicable`) also exist on `Collapse`; so internally whenever both `Collapse` and `CollapsePrivate` are in scope, we need to disambiguate between them using the `<Type as Trait>::method(args...)` syntax, which is fine I guess. That will **never** be necessary for users code, however, as they cannot import `CollapsePrivate` into their namespace.  So if you're worried about users needing to write `<Type as Trait>::method(args...)`, don't be. It's only internal code that will have to deal with that from now on (now that we've removed the `_` and made the trait method names the same across both `Collapse` and `CollapsePrivate`.

Reply:
I think making them the same given that one will generally just forward to the other makes sense to me :)